apps/labs/posts/dataframe-group-by.md

```diff
@@ -0,0 +1,167 @@
+- select the column(s) you want to aggregate;
+- specify an elementary aggregation function.
+
+Let's try something else: "find the maximum value of 'c', where 'b' is greater than its mean, per
+group 'a'".
+
+Unfortunately, the pandas API has no way to express this, meaning
+that no library which copies pandas can truly optimise such an
+operation in the general case.
+
+We could do
+```python
+df.groupby('a').apply(
+    lambda df: df[df['b'] > df['b'].mean()]['c'].max()
+)
+```
+
+However, that uses a Python `lambda` function and so is generally going to be inefficient.
+
+Another solution could be to do:
+```python
+df[df['b'] > df.groupby('a')['b'].transform('mean')].groupby('a')['c'].max()
+```
+This isn't too bad, but it involves doing two group-bys, and so is at least twice as slow as it could
+be.
+
+Finally, can rely on `GroupBy` caching its groups, in-place mutation of the original dataframe, and the
+fact that `'max'` skips over missing values, to write:
+```python
+gb = df.groupby("a")
+mask = df["b"] > gb["b"].transform("mean")
+df["result"] = np.where(mask, df["c"], np.nan)
+gb["result"].max()
+```
+This works, but it involved quite some manual optimisation, and is definitely not a general solution.
+
+Surely it's possible to do better?
+
+## Non-elementary group-bys in Polars
+
+The Polars API lets us pass expressions to `GroupBy.agg`. So long as you can express your aggregation as
+an expression, you can use it in a group-by setting. In this case, we can express "the maximum value
+of 'c' where 'b' is greater than its mean" as
+```python
+pl.col('c').filter(pl.col('b') > pl.mean('b')).max()
+```
+Then, all we need to do is pass this expression to `GroupBy.agg`:
+
+```python
+df.group_by('a').agg(
+    pl.col('c').filter(pl.col('b') > pl.mean('b')).max()
+)
+```
+Wonderful! We could express the operation cleanly using the library's own API, meaning that any implementation
+of the Polars API (not necessarily Polars itself) has the possibility to evaluate this efficiently.
+
+## Conclusion, and a plea to future dataframe authors
+
+We've learned about the group-by operation, elementary aggregations in both pandas and Polars, and how
+Polars' syntax enables optimisation of non-elementary aggregations.
+
+pandas is a wonderful tool which solves a lot of real problems for a lot of real people.
+But please, stop copying its API. No matter how much effort you put into
+your implementation, if your API is limited and can't express non-elementary group-by aggregations,
+then you're going to run into a wall at some point. Good luck reverse-parsing the bytecode of a user's
+Python lambda function.
+
+On the other hand, if you innovate on the API side, you can enable new possibilities. There's a reason
+that
+
+> I came for the speed, but stayed for the syntax
+
+is a common refrain among Polars users.
+
+There may be a more general lesson here: if you have the courage to do things differently, you may be rewarded.
```

Comment:
This line doesn't really connect for me with anything else in the blog post, seems to come out of nowhere.